File: elixir/lib/day1.ex
defmodule Day1 do
  # This function is equivalent to the Clojure:
  # (defn parse-input [input]
  #   (->> (str/split-lines input)
  #        (map #(mapv parse-long (str/split % #"\s+")))))
  def parse_input(input) do
    input
    # String.split("\n") is like str/split-lines
    |> String.split("\n", trim: true)
    # Enum.map is like Clojure's map
    # The fn -> end block is like Clojure's #(...)
    |> Enum.map(fn line ->
      line
      # ~r/\s+/ is a regex like #"\s+" in Clojure
      |> String.split(~r/\s+/)
      # &String.to_integer/1 is like parse-long
      # The & syntax is like Clojure's #()
      |> Enum.map(&String.to_integer/1)
    end)
  end

  # Equivalent to Clojure:
  # (defn solve-part1 [input]
  #   (let [data (parse-input input)
  #         pairs (map vector 
  #                   (sort (map first data))
  #                   (sort (map second data)))]
  #     (reduce + (map #(Math/abs (apply - %)) pairs))))
  def solve_part1(input) do
    data = parse_input(input)
    # List.first/1 is like Clojure's first
    # Enum.sort is like Clojure's sort
    firsts = data |> Enum.map(&List.first/1) |> Enum.sort()
    # List.last/1 is like Clojure's second (in this case, since we know there are 2 elements)
    seconds = data |> Enum.map(&List.last/1) |> Enum.sort()
    
    # Enum.zip is like Clojure's (map vector xs ys)
    Enum.zip(firsts, seconds)
    # The {a, b} pattern match is like Clojure's destructuring
    # abs(a - b) is like (Math/abs (apply - [a b]))
    |> Enum.map(fn {a, b} -> abs(a - b) end)
    # Enum.sum is like (reduce +)
    |> Enum.sum()
  end

  # Equivalent to Clojure:
  # (defn solve-part2 [input]
  #   (let [data (parse-input input)]
  #     (->> (map first data)
  #          (map #(* % (count (filter #{%} (map second data)))))
  #          (reduce +))))
  def solve_part2(input) do
    data = parse_input(input)
    firsts = Enum.map(data, &List.first/1)
    seconds = Enum.map(data, &List.last/1)
    
    firsts
    |> Enum.map(fn x -> 
      # Enum.count with a predicate is like (count (filter pred coll))
      # &(&1 == x) is like #(= % x) in Clojure
      # The whole expression is like (* x (count (filter #{x} seconds)))
      x * Enum.count(seconds, &(&1 == x))
    end)
    |> Enum.sum()
  end

  # Equivalent to Clojure:
  # (defn solve [] (solve-part1 (slurp (io/resource "day1.txt"))))
  def solve do
    # File.read! is like slurp
    # We use priv/ directory instead of resources in Elixir
    File.read!("priv/day1.txt")
    |> solve_part1()
  end

  # Equivalent to:
  # (defn solve2 [] (solve-part2 (slurp (io/resource "day1.txt"))))
  def solve2 do
    File.read!("priv/day1.txt")
    |> solve_part2()
  end
end

File: python/day1.py
"""
This module solves Day 1 of Advent of Code 2024.
Translated from Clojure to Python with explanatory comments for Clojure developers.
"""

from pathlib import Path
from typing import List, Tuple

def parse_input(input_str: str) -> List[List[int]]:
    """
    Parses the input string into a list of number pairs.
    
    In Clojure this was:
    (->> (str/split-lines input)
         (map #(mapv parse-long (str/split % #"\s+"))))
    
    The ->> is a thread-last macro that pipes the result of each expression
    as the last argument to the next expression. In Python we write this
    more imperatively.
    """
    # Split lines is like Clojure's str/split-lines
    lines = input_str.split('\n')
    
    # For each line, split on whitespace and convert to integers
    # This replaces Clojure's (map #(mapv parse-long (str/split % #"\s+")))
    # mapv in Clojure returns a vector instead of a lazy sequence
    return [list(map(int, line.split())) for line in lines]

def solve_part1(input_str: str) -> int:
    """
    Solves part 1 of the puzzle.
    
    In Clojure this was:
    (let [data (parse-input input)
          pairs (map vector 
                    (sort (map first data))
                    (sort (map second data)))]
      (reduce + (map #(Math/abs (apply - %)) pairs)))
    
    The let binding in Clojure creates local bindings. In Python we just
    use regular variable assignment.
    """
    data = parse_input(input_str)
    
    # Extract and sort first and second elements separately
    # In Clojure (map first data) gets first elements, (map second data) gets second elements
    firsts = sorted(row[0] for row in data)
    seconds = sorted(row[1] for row in data)
    
    # Create pairs using zip (similar to Clojure's map vector)
    pairs = list(zip(firsts, seconds))
    
    # Calculate absolute differences and sum them
    # Replaces (reduce + (map #(Math/abs (apply - %)) pairs))
    return sum(abs(a - b) for a, b in pairs)

def solve_part2(input_str: str) -> int:
    """
    Solves part 2 of the puzzle.
    
    In Clojure this was:
    (let [data (parse-input input)]
      (->> (map first data)
           (map #(* % (count (filter #{%} (map second data)))))
           (reduce +)))
    
    The ->> macro in Clojure threads the result through the expressions.
    In Python we write this more directly.
    """
    data = parse_input(input_str)
    
    # Get all first elements
    firsts = [row[0] for row in data]
    # Get all second elements
    seconds = [row[1] for row in data]
    
    # For each first number, multiply it by how many times it appears in seconds
    # In Clojure, #{%} creates a set with one element for filtering
    result = sum(x * seconds.count(x) for x in firsts)
    return result

def solve() -> int:
    """
    Solves part 1 using input from day1.txt.
    
    In Clojure this used:
    (slurp (io/resource "day1.txt"))
    
    Python doesn't have a direct equivalent to Clojure's io/resource,
    so we use pathlib to handle paths relative to this file.
    """
    input_path = Path(__file__).parent / "day1.txt"
    with open(input_path) as f:
        return solve_part1(f.read())

def solve2() -> int:
    """
    Solves part 2 using input from day1.txt.
    Similar to solve() but uses solve_part2.
    """
    input_path = Path(__file__).parent / "day1.txt"
    with open(input_path) as f:
        return solve_part2(f.read())

if __name__ == "__main__":
    print(f"Part 1: {solve()}")
    print(f"Part 2: {solve2()}")
